saved_replies: Add new feature for saved replies.

Vector73 opened the pull request:
<!-- Describe your pull request here.-->
Adds a new feature for adding saved replies.

Fixes: #31227<!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->


<details><summary>Screenshots</summary>
<p>

![Screenshot 2024-09-14 165827](https://github.com/user-attachments/assets/e3801270-c99f-48e8-b1d9-8503e2fd843c)
![Screenshot 2024-09-14 165947](https://github.com/user-attachments/assets/0833a0ef-6c67-44d0-a14e-0eea730c36c6)




</p>
</details> 
<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

roanster007 commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,23 @@ format used by the Zulip server that they are interacting with.
+**Feature level 284**
+
+* [`GET /events`](/api/get-events): An event with `type: "saved_replies"` is
+  sent to the current user when a saved reply is created or deleted.
+
+* [`POST /register`](/api/register-queue): Added a new `saved_replies` field
```
New line between 2 changes may not be needed.

roanster007 commented on web/src/saved_replies.ts:
```diff
@@ -0,0 +1,157 @@
+const ADD_SAVED_REPLY_OPTION_ID = -1;
+let my_saved_replies: SavedReplies[] = [];
+
+export function set_saved_replies(saved_replies: SavedReplies[]): void {
+    my_saved_replies = saved_replies;
+    my_saved_replies.sort((a, b) => a.title.toLowerCase().localeCompare(b.title.toLowerCase()));
```
It may be fine to exempt `saved_replies.ts` from test coverage, but I guess it would be better if we write some tests that verify the state of the `my_saved_replies` after performing each of the functions like  `remove_saved_reply`, `add_saved_reply`, etc.

I guess you could have a look at `stream_data.test.js` as an example for `get_options_for_dropdown_widget` test?

Vector73 replied:
Added tests for `add_saved_reply`, `remove_saved_reply` and `get_options_for_dropdown_widget`.

roanster007 commented on zerver/lib/events.py:
```diff
@@ -204,6 +205,12 @@ def fetch_initial_state_data(
         state["max_message_id"] = max_message_id_for_user(user_profile)
 
+    if want("saved_replies"):
+        if user_profile is None:
+            state["saved_replies"] = []
+        saved_replies = SavedReply.objects.filter(user_profile=user_profile)
+        state["saved_replies"] = [saved_reply.to_dict() for saved_reply in saved_replies]
```
There is an `else` missing I guess?

Vector73 replied:
Oh, missed that. Nice catch.

roanster007 commented on web/src/dropdown_widget.ts:
```diff
@@ -398,7 +398,7 @@ export class DropdownWidget {
 
                 // Click on item.
-                $popper.one("click", ".list-item", (event) => {
+                $popper.on("click", ".list-item", (event) => {
```
What's the reason for this change?

Vector73 replied:
The saved replies can be deleted by clicking on the trash icon in dropdown. But with `popper.one` only one click is allowed in the dropdown. So I changed it to `popper.on` to allow multiple deletions without dropdown closing.

Vector73 commented on zerver/lib/saved_replies.py:
```diff
@@ -0,0 +1,56 @@
+        "op": "add",
+        "saved_reply": saved_reply.to_dict(),
+    }
+    send_event(user_profile.realm, event, [user_profile.id])
```
Should I change it to `send_event_on_commit` then?

roanster007 replied:
Ya, update it.

roanster007 commented:
![image](https://github.com/user-attachments/assets/61ccd790-f35a-40e1-a216-ae5994f391b5)
`Add a new saved reply` is unresponsive to me.. could you check it out

Vector73 commented:
It should work now.

roanster007 commented:
Creating a saved reply, and then removing it , and then searching for without closing is showing results of the deleted saved reply.

![brave_t0nv9iLzut](https://github.com/user-attachments/assets/784de823-6501-4308-8d70-7fd5b481fa9a)

roanster007 commented on web/src/saved_replies.ts:
```diff
@@ -0,0 +1,157 @@
+
+export function add_saved_reply(saved_reply: SavedReplies): void {
+    my_saved_replies.push(saved_reply);
+    my_saved_replies.sort((a, b) => a.title.toLowerCase().localeCompare(b.title.toLowerCase()));
```
```suggestion
    my_saved_replies.sort((a, b) => util.strcmp(a.name.toLowerCase(), b.name.toLowerCase()));
```

roanster007 commented on web/src/saved_replies.ts:
```diff
@@ -0,0 +1,157 @@
+    }
+
+    dropdown.hide();
+    const current_value = widget.current_value;
```
`widget` is a generic name, make it something like `saved_replies_widget`

Vector73 replied:
But there are other places where only `dropdown` is used as argument in `item_click_callback` like in `settings_stream.ts`.

roanster007 replied:
Ok, I guess you can leave it like that then, seems like it is `dropdown` everywhere.

Vector73 commented:
> Creating a saved reply, and then removing it , and then searching for without closing is showing results of the deleted saved reply.

Fixed this bug. Nice catch!

roanster007 commented on web/src/saved_replies.ts:
```diff
@@ -0,0 +1,161 @@
+    const saved_replies = my_saved_replies.map((saved_reply) => ({
+        unique_id: saved_reply.id,
+        name: saved_reply.title,
+        description: saved_reply.rendered_content.replace(/^(?:<p>)?(.*?)(?:<\/p>)?$/, "$1"),
```
Any specific reason why the `<p>` tags being removed from the rendered_content?

Vector73 replied:
The `<p>` is removed here to show `rendered_content` in a single line.

Vector73 replied:
The `<p>` is removed here to show `rendered_content` in a single line in the dropdown.

roanster007 replied:
Ok, that makes sense.

timabbott replied:
Hacky HTML manipulation like this can be dangerous. I'd prefer that we use CSS to manage how the rendered content gets displayed.

roanster007 commented on zerver/tests/test_saved_replies.py:
```diff
@@ -0,0 +1,88 @@
+        result = self.client_get(
+            "/json/saved_replies",
+        )
+        saved_reply = SavedReply.objects.filter(user_profile=user).order_by("id")[0]
```
I guess you need to be using `do_get_saved_replies` rather than filtering using the object to test `get_saved_replies`. And also, make sure to have 2 saved replies before  testing this.

roanster007 commented on web/tests/dispatch.test.js:
```diff
@@ -177,6 +178,15 @@ run_test("alert_words", ({override}) => {
 });
 
+run_test("saved_replies", ({override}) => {
+    override(saved_replies, "add_saved_reply", noop);
+    override(saved_replies, "remove_saved_reply", noop);
+    const add_event = event_fixtures.saved_replies__add;
+    dispatch(add_event);
+    const remove_event = event_fixtures.saved_replies__remove;
+    dispatch(remove_event);
```
```suggestion
    assert.equal(saved_replies.get_saved_replies().length, 0)
   
    const add_event = event_fixtures.saved_replies__add;
    dispatch(add_event);
    assert.equal(saved_replies.get_saved_replies().length, 1)
    
    const remove_event = event_fixtures.saved_replies__remove;
    dispatch(remove_event);
    assert.equal(saved_replies.get_saved_replies().length, 0)
```

roanster007 replied:
You need to test something after dispatching the event. Hence, the change.

Vector73 replied:
Updated to make it consistent with other tests:

```js
const add_event = event_fixtures.saved_replies__add;
    {
        const stub = make_stub();
        override(saved_replies, "add_saved_reply", stub.f);

        dispatch(add_event);
        console.log(stub);
        assert.equal(stub.num_calls, 1);
        assert_same(stub.get_args("event").event, add_event.saved_reply);
    }

    const remove_event = event_fixtures.saved_replies__remove;
    {
        const stub = make_stub();
        override(saved_replies, "remove_saved_reply", stub.f);

        dispatch(remove_event);
        assert.equal(stub.num_calls, 1);
        assert_same(stub.get_args("event").event, remove_event.saved_reply_id);
    }
```

roanster007 commented on web/tests/dispatch.test.js:
```diff
@@ -49,6 +49,7 @@ const realm_icon = mock_esm("../src/realm_icon");
 const realm_logo = mock_esm("../src/realm_logo");
 const realm_playground = mock_esm("../src/realm_playground");
 const reload = mock_esm("../src/reload");
+const saved_replies = mock_esm("../src/saved_replies");
```
```suggestion
const saved_replies = zrequire("../src/saved_replies");
```

roanster007 replied:
Using `zrequire` was actually part of my suggestion in https://github.com/zulip/zulip/pull/31359#discussion_r1734034112. So basically using `zrequire` would allow you to maintain the module and its functions as they are -- So like in this case, you can actually store and get things from `saved_replies` data structure... Just like how `message_store` is being tested in here.

`mock_esm` basically mocks the module so that you can override functions as you wish... I feel verifying the contents of data structure from the event dispatch would make more sense than verifying just whether the function is called.

timabbott replied:
This module uses `mock_esm` for basically everything, probably it's reasonable to preserve that model. But I think the `saved_replies/saved_replies_ui` split would help here.

roanster007 commented on zproject/urls.py:
```diff
@@ -329,6 +330,8 @@
     # Endpoints for syncing drafts.
     rest_path("drafts", GET=fetch_drafts, POST=create_drafts),
     rest_path("drafts/<int:draft_id>", PATCH=edit_draft, DELETE=delete_draft),
+    rest_path("saved_replies", GET=get_saved_replies, POST=create_saved_reply),
+    rest_path("saved_replies/<int:saved_reply_id>", DELETE=delete_saved_reply),
```
Make sure to add a separate comment to your changes from `# Endpoints for syncing drafts.`

roanster007 commented:
I guess we should also deal with duplicate saved replies entries... I am not sure whether to reject entries with same title and content, or only same content .... But I feel it wouldn't make sense to have duplicate entries.

Vector73 commented on zerver/lib/saved_replies.py:
```diff
@@ -0,0 +1,56 @@
+    content: str,
+    user_profile: UserProfile,
+) -> SavedReply:
+    title = title.strip()[: SavedReply.MAX_TITLE_LENGTH]
+    content = content.strip()
```
Asked on [CZO](https://chat.zulip.org/#narrow/stream/378-api-design/topic/saved.20replies/near/1932517) for deciding the correct behaviour.

timabbott replied:
Yeah I'd recommend doing the `strip` step in the caller, and rejecting requests where either in the empty string after stripping.

(Generally as a naming convention, `do_` functions are those that can't fail).

Vector73 commented on zerver/lib/saved_replies.py:
```diff
@@ -0,0 +1,56 @@
+    try:
+        saved_reply = SavedReply.objects.get(id=saved_reply_id, user_profile=user_profile)
+    except SavedReply.DoesNotExist:
+        raise JsonableError(_("Saved reply does not exist."))
```
Updated that.

Vector73 commented:
> I guess we should also deal with duplicate saved replies entries... I am not sure whether to reject entries with same title and content, or only same content .... But I feel it wouldn't make sense to have duplicate entries?

This has already been discussed on [here](https://chat.zulip.org/#narrow/stream/378-api-design/topic/saved.20replies/near/1916036) and the consensus was that we don't want the saved replies to be unique.

roanster007 commented:
LGTM

timabbott commented on zerver/models/saved_replies.py:
```diff
@@ -0,0 +1,25 @@
+            "content": self.content,
+            "rendered_content": self.rendered_content,
+            "date_created": int(self.date_created.timestamp()),
+        }
```
I was thinking about whether we need any indexes, but probably the `user_profile` `ForeignKey` is the only thing that would end up being used. But maybe add a `realm` column? We've denormalized that immutable field into a decent number of other tables, and we'll need it if in the future we add shared saved replies.

Vector73 replied:
Added `realm` field.

timabbott commented on zerver/models/saved_replies.py:
```diff
@@ -0,0 +1,25 @@
+    user_profile = models.ForeignKey(UserProfile, on_delete=models.CASCADE)
+    title = models.CharField(max_length=MAX_TITLE_LENGTH)
+    content = models.TextField()
+    rendered_content = models.TextField()
```
Usually we have a `rendered_content_version` where we have `rendered_content`.

timabbott commented on zerver/lib/saved_replies.py:
```diff
@@ -0,0 +1,56 @@
+    # Used in the UI in dropdown for showing the rendered version of the content in a single line.
+    rendered_content = markdown_convert(
+        content.replace("\n", " "), message_realm=user_profile.realm, no_previews=True
+    ).rendered_content
```
Why are you replacing newlines with spaces here? I guess the motivation seems to be for fitting these into a limited area... Can you do some more extensive screenshots, maybe in the design thread for the feature, of various more complex syntax in a saved reply (bulleted lists / images / etc.). My guess is some of those will be corrupted and this is undesirable.

(We can manage space in other ways in the UI; simplest would be to just use CSS to cut off the region displayed)

timabbott replied:
I'm also not 100% sure we want to do Markdown rendering in the server, vs. having the web app render these when loading the view, like we do with Drafts.

timabbott commented on zerver/lib/events.py:
```diff
@@ -204,6 +205,13 @@ def fetch_initial_state_data(
+            state["saved_replies"] = []
+        else:
+            saved_replies = SavedReply.objects.filter(user_profile=user_profile)
+            state["saved_replies"] = [saved_reply.to_dict() for saved_reply in saved_replies]
```
We might need to specify an explicit `to_api_dict` function after my `models.py` comments are done to only include the fields that we want in the API.

timabbott replied:
Also this block should call `do_get_saved_replies` rather than duplicating it; that approach sets us up for success in future refactoring. The above comment still applies to the `do_get_saved_replies` implementation.

timabbott replied:
Except the above comment about `to_api_dict` is nonsense because `to_dict` is a function defined only for this purpose :). `to_api_dict` might be a nicer name though.

timabbott commented on zerver/lib/export.py:
```diff
@@ -160,6 +160,7 @@ class MessagePartial(TypedDict):
     "zerver_realmreactivationstatus",
     "zerver_realmuserdefault",
     "zerver_recipient",
+    "zerver_savedreply",
```
I expect this needs `import_realm.py` code as well, to renumber the `UserProfile` and future `Realm` foreign keys. You can test it by creating a `savedreply` object in the early part of `test_import_export.py` and having it diff the before/after database state.

Vector73 replied:
So, is this correct?:

```py
if "zerver_savedreply" in data:
        re_map_foreign_keys(data, "zerver_savedreply", "user_profile", related_table="user_profile")
        re_map_foreign_keys(data, "zerver_savedreply", "realm", related_table="realm")
        update_model_ids(SavedReply, data, "savedreply")
        bulk_import_model(data, SavedReply)
```

timabbott commented on zerver/tests/test_saved_replies.py:
```diff
@@ -0,0 +1,79 @@
+        user = self.example_user("hamlet")
+
+        SavedReply.objects.filter(user_profile=user).delete()
+        return user
```
I'm not totally sure why this helper would be needed; the test database won't come with any `SavedReply` objects unless we add them in `populate_db` ... which we might want to do, but only in the "dev environment only" section, not the section shared with the test suite.

And the database state gets reset at the end of each test using transactions.

Vector73 replied:
Removed.

timabbott commented on zerver/tests/test_saved_replies.py:
```diff
@@ -0,0 +1,79 @@
+        saved_reply = do_create_saved_reply(
+            "Welcome message", "**Welcome** to the organization.", user
+        )
+        return saved_reply.id
```
It's probably better for this helper to accept the user as a parameter.

timabbott commented on zerver/tests/test_saved_replies.py:
```diff
@@ -0,0 +1,79 @@
+            "/json/saved_replies",
+        )
+        response_dict = self.assert_json_success(result)
+        self.assertEqual(response_dict["saved_replies"], do_get_saved_replies(user))
```
I'm not sure this test is verifying much new. The other two tests should be using the API rather than `do_get_saved_replies` to verify their state, you might find it reasonable to just delete this test, as those other endpoints would then be verifying the get endpoint naturally. See https://zulip.readthedocs.io/en/stable/testing/philosophy.html#integration-testing-or-unit-testing for the discussion of this end-to-end preference.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -22221,6 +22421,34 @@ components:
+        content:
+          type: string
+          description: |
+            The content of the saved reply in text/markdown format.
```
I'd continue this with a new paragraph of the form "Clients should insert this content into a message when using a saved reply".

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -4970,6 +4970,67 @@ paths:
+                            - type: object
+                              additionalProperties: false
+                              description: |
+                                Event containing details of a newly created saved reply.
```
This one is missing `**Changes**: New in Zulip 10.0 (feature level 284).`

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -5714,6 +5775,134 @@ paths:
+      tags: ["saved_replies"]
+      summary: Create a saved reply
+      description: |
+        Create a new saved reply for the current users.
```
"for the current user".

timabbott commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,19 @@ format used by the Zulip server that they are interacting with.
+* [`GET /events`](/api/get-events): An event with `type: "saved_replies"` is
+  sent to the current user when a saved reply is created or deleted.
+* [`POST /register`](/api/register-queue): Added a new `saved_replies` field
+  for current user's saved replies.
```
Generally, we list `events` and `register` together on a single line; probably you should follow that convention.

Vector73 replied:
Changed it to --
```md
* [`GET /events`](/api/get-events), [`POST /register`](/api/register-queue):
  An event with `type: "saved_replies"` is sent to the current user when a
  saved reply is created or deleted.
```

timabbott commented on web/src/saved_replies.ts:
```diff
@@ -0,0 +1,161 @@
+
+export function remove_saved_reply(saved_reply_id: number): void {
+    my_saved_replies = my_saved_replies.filter((s) => s.id !== saved_reply_id);
+}
```
Our standard convention these days is to put the above block of code into `saved_replies.ts`, and the UI code below it into a new `saved_replies_ui.ts`; this separation often ends up being helpful for avoiding import cycles. (The former module imports almost nothing, yet needs to be imported by code paths like `server_events_dispatch.ts`)

timabbott replied:
Actually `get_options_for_dropdown_widget` could probably go into the data module side of that divide too, since it's a nice, pure, testable function.

timabbott commented on web/src/saved_replies.ts:
```diff
@@ -0,0 +1,161 @@
+            submit_create_saved_reply_form();
+        }
+    });
+}
```
Hmm, does `dialog_widget` not provide a way to configure enter-submits without special code like this?

Vector73 replied:
Yeah, I noticed it does when `form_id` is provided in the constructor. Updated it now.

timabbott commented on web/src/saved_replies.ts:
```diff
@@ -0,0 +1,161 @@
+        const id = $(event.currentTarget).attr("data-unique-id");
+        if (id === undefined) {
+            return;
+        }
```
This should probably be `assert id !== undefined`; it's be a software bug that should be raised if this was ever not present.

timabbott commented on web/src/saved_replies.ts:
```diff
@@ -0,0 +1,161 @@
+                const options = get_options_for_dropdown_widget();
+                widget.list_widget?.replace_list_data(
+                    options.filter((option) => option.unique_id !== saved_reply_id),
+                );
```
Generally we do this sort of rerender via `server_events_dispatch` triggers. If you're not doing that, at least add a comment here noting that would be preferred to properly handle live-update in the unlikely event that another open browser window for this user deleted the thing.

Vector73 replied:
Updated the code to rerender the dropdown via `server_events_dispatch` when saved reply is added and deleted.

timabbott commented on web/src/saved_replies.ts:
```diff
@@ -0,0 +1,161 @@
+            },
+        });
+        return;
+    }
```
From a readability standpoint, probably you should make the deletion code a function rather than a big `if` block here.

timabbott commented on web/src/saved_replies.ts:
```diff
@@ -0,0 +1,161 @@
+        )?.content;
+        assert(content !== undefined);
+        const $textarea = $<HTMLTextAreaElement>("textarea#compose-textarea");
+        compose_ui.insert_and_scroll_into_view(content, $textarea);
```
Have we discussed the exact semantics around whether it should be padded with spaces or newlines here? Might be worth discussing on the design thread. GitHub's similar feature seems to insert a newline before the saved reply, and not after; I have no reason to think that's better than a space in the chat context.

Vector73 replied:
Used `insert_syntax_and_focus`.

timabbott commented on web/src/saved_replies.ts:
```diff
@@ -0,0 +1,161 @@
+        focus_target_on_hidden: false,
+        prefer_top_start_placement: true,
+        tippy_props: {
+            offset: [-10, 5],
```
Can you add a comment documenting these numbers?

Vector73 replied:
They were copied from compose recipient dropdown. Changed them now and added a comment:
```
// Using -100 as x offset makes saved reply icon be in the center
// of the dropdown widget and 5 as y offset is what we use in compose
// recipient dropdown widget.
```

timabbott commented on web/templates/dropdown_list.hbs:
```diff
@@ -1,18 +1,29 @@
     <li class="list-item" role="presentation" data-unique-id="{{unique_id}}" data-name="{{name}}" tabindex="0">
         {{#if description}}
-        <a class="dropdown-list-item-common-styles">
+        <span class="dropdown-list-item-common-styles">
```
The specific bit of refactoring of this from an `a` to a `span` probably belongs in its own commit, since that's the main thing here that could break other functionality. Probably you should just do it in its own commit in the prep PR tweaking the dropdown widget component that I suggested in another comment.

Vector73 replied:
Not needed now.

Vector73 replied:
Opened #31488 to replace `popper.one` to `popper.on`.

timabbott commented on web/templates/dropdown_list.hbs:
```diff
@@ -1,18 +1,29 @@
+                        <span id="delete-saved-reply-id" class="delete-saved-reply">
+                            <i class="fa fa-trash-o delete-saved-reply"></i>
+                        </span>
+                    {{/if}}
```
A few notes on this:
- Don't use the same `id` for an element that will have multiple copies of it exist; it's invalid HTML to have duplicate IDs. You also have the same class on both the wrapping `span` element and the inner `i` element. `delete-saved-reply-icon` for the inner one and `delete-saved-reply` on the outer one would be a fine class naming scheme.
- That all said, we also want to have this file be as generic as possible, not filled with customizations for specific features. I think you could just rename `is_saved_reply` to `has_delete_icon`, call the classes `dropdown-list-delete` / `dropdown-list-delete-icon`, and achieve that goal with pretty minimal effort.

timabbott commented:
I got in a pretty complete review of this; I think there's still some things to figure out with how we want to handle the `rendered_markdown` display logic, which you might need to get some help with in #frontend. But I think the above is most of the detail-level feedback that this should need before integration.

Vector73 commented:
Updates made in the recent changes:
- Added `realm` field to `SavedReply` table.
- Removed `rendered_content` from `SavedReply` table as raw content is used in the saved reply dropdown rather than the markdown version - decided [here](https://chat.zulip.org/#narrow/stream/101-design/topic/saved.20replies/near/1934394).
- Split `saved_replies.ts` in `saved_replies.ts` (for updating and storing saved reply data) and `saved_replies_ui.ts` (for UI logic).
- Rerender dropdown widget via `server_events_dispatch.js` whenever saved reply is added or deleted to show changes across multiple browser windows.
- Used `insert_syntax_and_focus` instead of `insert_and_scroll_into_view` for adding content to compose box.
- Moved saved reply button to 3-dot menu - https://chat.zulip.org/#narrow/stream/101-design/topic/saved.20replies/near/1934396.
- Opened prep PR for replacing `$popper.one` to `$popper.on` in `dropdown_widget.ts` to ensure correct behaviour for deleting saved replies -- #31488.
- Updated `import_realm.py`, `zulip.yaml` and some tests related to saved replies.

Vector73 commented:
Renamed saved replies to saved snippets as decided [here](https://chat.zulip.org/#narrow/stream/101-design/topic/saved.20replies/near/1941637).

timabbott commented on tools/test-js-with-node:
```diff
@@ -195,6 +195,8 @@ EXEMPT_FILES = make_set(
         "web/src/resize_handler.ts",
         "web/src/rows.ts",
+        "web/src/saved_snippets.ts",
+        "web/src/saved_snippets_ui.ts",
```
Usually I'd expect it to be easy to have 100% test coverage on the non-UI module.

timabbott commented on web/tests/saved_replies.test.js:
```diff
@@ -0,0 +1,112 @@
+    my_saved_snippets = saved_snippets.get_saved_snippets();
+    assert.equal(my_saved_snippets.length, 2);
+    assert.equal(my_saved_snippets[0].id, 2);
+});
```
This file still has `saved_replies` in the name.

timabbott commented on zerver/lib/saved_snippets.py:
```diff
@@ -0,0 +1,49 @@
+    }
+    send_event_on_commit(user_profile.realm, event, [user_profile.id])
+
+    return saved_snippet
```
This should use `@transaction.atomic(durable=True)`, and probably have a `RealmAuditLog` entry; we've usually regretted adding new data without using that.

Also, this file should probably be `zerver/actions/saved_snippets.py`, given the type of functions you have in here.

Vector73 replied:
I have moved the file to `actions` folder and added a `RealmAuditLog` entry for add function.

timabbott commented on zerver/models/saved_snippets.py:
```diff
@@ -0,0 +1,25 @@
+
+    realm = models.ForeignKey(Realm, on_delete=models.CASCADE)
+    user_profile = models.ForeignKey(UserProfile, on_delete=models.CASCADE)
+    title = models.CharField(max_length=MAX_TITLE_LENGTH)
```
This should probably be a `TextField`; the benefit if that it's a lot less messy to change the `max_length` later.

timabbott commented on zerver/views/saved_snippets.py:
```diff
@@ -0,0 +1,47 @@
+    if title == "":
+        raise JsonableError(_("Title cannot be empty."))
+    if content == "":
+        raise JsonableError(_("Content cannot be empty."))
```
This should check for title being too long, and reject in that case. `content` should also be limited to `MAX_MESSAGE_LENGTH`. Use `StringConstraints` from Pydantic to do all the enforcement if possible rather than writing custom code like this, which in particular requires translators to translate additional strings.

timabbott replied:
Also, posted https://chat.zulip.org/#narrow/stream/378-api-design/topic/saved.20replies/near/1942386 with some more questions.

Vector73 replied:
Changed `MAX_TITLE_LENGTH` to `60` and the request is rejected if the title's length exceeds that.

timabbott commented on zerver/tests/test_openapi.py:
```diff
@@ -902,6 +902,7 @@ def test_attributes(self) -> None:
             "scheduled_messages",
             "mobile",
             "invites",
+            "saved_snippets",
```
This is I think not necessary, if we're putting them in the `drafts` or `users` section?

Vector73 replied:
Added them to `drafts` section.

timabbott commented on zerver/lib/import_realm.py:
```diff
@@ -1425,6 +1426,14 @@ def do_import_realm(import_dir: Path, subdomain: str, processes: int = 1) -> Rea
+        )
+        re_map_foreign_keys(data, "zerver_savedsnippet", "realm", related_table="realm")
+        update_model_ids(SavedSnippet, data, "savedsnippet")
+        bulk_import_model(data, SavedSnippet)
```
How have you tested this code? Usually we extend the data in the existing `test_import_export` tests to creates one object and verify it survives the data export/import process.

Vector73 replied:
I tried adding it to `test_import_export` in `SingleUserExportTest` and creating a `checker` for `zerver_savedsnippet` but I am getting an error that there is no `zerver_savedsnippet` key in `user.json`.
```sh
Traceback (most recent call last):
  File "/home/vector/zulip/zerver/tests/test_import_export.py", line 2289, in test_user_data
    f(user[table_name])
KeyError: 'zerver_savedsnippet'
```

timabbott replied:
OK, that probably means this code has a bug.

timabbott commented on web/src/saved_snippets.ts:
```diff
@@ -0,0 +1,51 @@
+};
+
+export const ADD_SAVED_SNIPPET_OPTION_ID = -1;
+let my_saved_snippets: SavedSnippet[] = [];
```
Use a map data structure keyed by the IDs; the array format is only for network transit. Half of your queries are things that would be an O(1) lookup in a map.

(At least one `get_saved_snippets` caller is iterating through the list to find a single item by ID, and `snippet` is also very strange)

Vector73 replied:
OK, used map instead of an array for storing saved snippets on the client side and changed `get_saved_snippets` function to `get_saved_snippet_by_id`.

timabbott commented on web/src/saved_snippets.ts:
```diff
@@ -0,0 +1,51 @@
+
+export function add_saved_snippet(saved_snippet: SavedSnippet): void {
+    my_saved_snippets.push(saved_snippet);
+    my_saved_snippets.sort((a, b) => util.strcmp(a.title.toLowerCase(), b.title.toLowerCase()));
```
Probably this sort belongs in `get_options_for_dropdown_widget`.

While the data sizes are small enough that it likely doesn't matter performance-wise, one should always use the right data structures for what you're doing by default.

timabbott commented on web/src/saved_snippets_ui.ts:
```diff
@@ -0,0 +1,110 @@
+            .get_saved_snippets()
+            .find((saved_snippet) => saved_snippet.id === current_value)?.content;
+        assert(content !== undefined);
+        const $textarea = $<HTMLTextAreaElement>("textarea#compose-textarea");
```
When we make these available in the message-edit form as well, we'll need to make this a parameter to this function. Might be worth doing that plumbing now just to avoid having to do more refactoring later, but it's not strictly required.

Vector73 replied:
Did you mean to say that I should make `saved_snippet_id` a parameter to `get_saved_snippets`?

timabbott replied:
Yeah.

timabbott commented on web/src/saved_snippets_ui.ts:
```diff
@@ -0,0 +1,110 @@
+        },
+    });
+    saved_snippet_dropdown_widget.setup();
+};
```
I don't think you need to call this function from `ui_init.ts`; you only need this dropdown widget if someone clicks the relevant element in the compose UI, and doing it this way wastes time during web app startup.

The `send_later_popover` code should be the caller for this function, which probably should have a name more like `show`, with care taken to clean up the object when the popover is hidden.

Vector73 replied:
There is a error that I am encountering when initializing dropdown widget in `compose_send_menu_popover` -- `saved_snippet_dropdown_widget.list_widget` is getting `undefined` for some reason. I will try to resolve it but is it necessary to avoid initializing at startup? The `compose_recipient_dropdown` is also initialized in `ui_init.ts`.

timabbott replied:
I think we can try to do this in a smaller follow-up PR. It's not essential, just a small initial loading time optimization.

timabbott commented on web/src/saved_snippets_ui.ts:
```diff
@@ -0,0 +1,110 @@
+}
+
+function delete_saved_snippet(saved_snippet_id: string | undefined): void {
+    assert(saved_snippet_id !== undefined);
```
A more standard type design would be for this to require a string and have the caller do the assert.

timabbott commented:
I think that's a complete set of technical feedback for getting this PR into a mergeable state. Thanks for all the effort on this @Vector73!

alya commented:
Nice! Some feedback (can be follow-ups after this is merged but before it goes to production):

1. "Add a new saved snippet" is too bold. It should have this font weight:
![Screenshot 2024-09-24 at 15 00 00@2x](https://github.com/user-attachments/assets/3371960f-7554-4bdf-895e-aebfaf05c5eb)

1. When you open the modal to add a new snippet, your cursor should immediately be in the Title field.
2. The vertical spacing between the snippet title and content is much too wide, making it hard to parse what title snippet is associated with.

alya commented:
@timabbott up to you if you want to merge before fixing this one, but it's a clear bug: If you choose a snippet with the keyboard rather than the mouse, the compose menu fails to close like it ought.

alya commented:
I think it's a bit easy to misclick on the trash icon. Probably worth adding a confirmation modal. I can write up a brief spec assuming @timabbott confirms that this isn't too tricky from a technical perspective.

alya commented:
We should also consider scooting the "delete" button to the left a bit. On my Mac, it overlaps with the scrollbar:

![Screenshot 2024-09-24 at 15 11 19@2x](https://github.com/user-attachments/assets/d5759949-8899-4257-81df-980d127ddf75)

alya commented:
That's all I've got for now!

timabbott commented on version.py:
```diff
@@ -34,7 +34,8 @@
 
-API_FEATURE_LEVEL = 296  # Last bumped for `editable_by_user` custom profile field setting.
+
+API_FEATURE_LEVEL = 297  # Last bumped for saved_snippets
```
Don't add an extra blank line here :)

timabbott commented on zerver/actions/saved_snippets.py:
```diff
@@ -0,0 +1,61 @@
+        modified_user=user_profile,
+        event_type=AuditLogEventType.SAVED_SNIPPET_CREATED,
+        event_time=timezone_now(),
+    )
```
This doesn't encode the `saved_snippet_id`, so is mostly useless.

timabbott commented on zerver/models/realm_audit_logs.py:
```diff
@@ -129,6 +129,8 @@ class AuditLogEventType(IntEnum):
     REMOTE_REALM_LOCALLY_DELETED = 20003
     REMOTE_REALM_LOCALLY_DELETED_RESTORED = 20004
 
+    SAVED_SNIPPET_CREATED = 25001
```
This is not the right part of the file or numbering space for this feature.

timabbott replied:
Made this change to clean up this part of the system:

``` diff
diff --git a/zerver/actions/saved_snippets.py b/zerver/actions/saved_snippets.py
index cca10d3c4d..59ca88467f 100644
--- a/zerver/actions/saved_snippets.py
+++ b/zerver/actions/saved_snippets.py
@@ -29,6 +29,7 @@ def do_create_saved_snippet(
         modified_user=user_profile,
         event_type=AuditLogEventType.SAVED_SNIPPET_CREATED,
         event_time=timezone_now(),
+        extra_data={"saved_snippet_id": saved_snippet.id},
     )
 
     event = {
diff --git a/zerver/models/realm_audit_logs.py b/zerver/models/realm_audit_logs.py
index 6864805036..f6c6c534a3 100644
--- a/zerver/models/realm_audit_logs.py
+++ b/zerver/models/realm_audit_logs.py
@@ -108,6 +108,8 @@ class AuditLogEventType(IntEnum):
     USER_GROUP_GROUP_BASED_SETTING_CHANGED = 722
     USER_GROUP_DEACTIVATED = 723
 
+    SAVED_SNIPPET_CREATED = 800
+
     # The following values are only for remote server/realm logs.
     # Values should be exactly 10000 greater than the corresponding
     # value used for the same purpose in realm audit logs (e.g.,
@@ -129,8 +131,6 @@ class AuditLogEventType(IntEnum):
     REMOTE_REALM_LOCALLY_DELETED = 20003
     REMOTE_REALM_LOCALLY_DELETED_RESTORED = 20004
 
-    SAVED_SNIPPET_CREATED = 25001
-
 
 class AbstractRealmAuditLog(models.Model):
     """Defines fields common to RealmAuditLog and RemoteRealmAuditLog."""
```

timabbott commented:
I extracted and merged the backend of this as 9e4e85e1402378bf971b87c5f6223d7775db5fba, after the few small fixes noted above, which should make rebasing the remainder easy while we work through those UI tweaks.

sayamsamal commented:
> Nice! Some feedback (can be follow-ups after this is merged but before it goes to production):
> 
> 4. The trash can button needs a hover behavior. @sayamsamal do you have a suggestion on how best to do that right now?

We can use these colors to add the hover behavior to the trash can:
Normal State: `color-mix(in oklch, hsl(359deg 93% 39%) 70%, transparent);`
Hover State: `hsl(359deg 93% 39%);`
Active State: `color-mix(in oklch, hsl(359deg 93% 39%) 70%, transparent);`

Please also add a note that  `hsl(359deg 93% 39%)` corresponds to `var(--red-550)`. You'll find similar comments for `--color-copy-btn` in `web/styles/app_variables.css`.

Vector73 commented on web/src/dropdown_widget.ts:
```diff
@@ -145,7 +145,7 @@ export class DropdownWidget {
             `${this.widget_selector}, ${this.widget_wrapper_id}`,
             (e) => {
                 if (e.key === "Enter") {
-                    $(this.widget_selector).trigger("click");
+                    $(this.widget_selector)[0]?.click();
                     e.stopPropagation();
```
> Keyboard navigation doesn't open saved snippets from the menu -- pressing Enter with the option selected just makes the focus ring disapper.

I had to make this change to resolve this bug. Can it have any unexpected behaviour - using `click()` instead of `trigger("click")`?

timabbott replied:
This may be a pattern we want to avoid, because it makes migrating away from jQuery harder, but we do use it in the codebase, and it's not a huge deal.

Oddly, I can confirm it is required to make this work. I really don't understand how the difference between these two ways of triggering a click could possibly be important. 

Maybe I can merge this, but @amanagr can investigate and post in #frontend his findings.

Vector73 commented:
> 1. If you choose a snippet with the keyboard rather than the mouse, the compose menu fails to close like it ought.

I think this would need some custom logic `dropdown_widget` for closing the `send_later_popover` when a saved snippet is selected using `Enter` key. I am not sure if that would be the right solution though. @alya can we leave it to the current behaviour for now?

Vector73 commented:
> I think it's a bit easy to misclick on the trash icon. Probably worth adding a confirmation modal.

But if the user want to delete multiple snippets at a time, it would require a lot of clicks as opening the confirmation modal will close both dropdowns. @alya thoughts?

timabbott commented:
I don't think deleting a lot of snippets at once is going to be very common, so the confirmation modal strategy is probably a good one.

> I think this would need some custom logic dropdown_widget for closing the send_later_popover when a saved snippet is selected using Enter key. I am not sure if that would be the right solution though. @alya can we leave it to the current behaviour for now?

Start a thread in #frontend about tactics for this; probably @amanagr will be able to help you find a solution. This sounds like it looks broken enough that we will want to fix it before advertising the feature, so we might as well just fix it now.

alya commented:
Those screenshots look like an improvement; I can retest once all or most of the fixes are in.

Vector73 commented:
Fixed the `send_later_popover` not closing bug and added a confirmation modal for deleting saved snippets:

![Screenshot 2024-09-27 213923](https://github.com/user-attachments/assets/9b852fc9-ce41-4ee4-901d-c3d37017831f)

alya commented:
Testing this version, clicking on a snippet fails to close the send menu.

alya commented:
I haven't re-tested everything carefully, but looks good otherwise in light manual testing!

Vector73 commented:
> Testing this version, clicking on a snippet fails to close the send menu.

Fixed it now.

timabbott commented on web/styles/compose.css:
```diff
@@ -1588,6 +1588,51 @@ textarea.new_message_textarea {
+        margin-bottom: 20px;
+    }
+
+    & .content {
```
These would really benefit from better CSS class names, like `saved-snippet-title`. Generally you want every class name to be globally unique in the app, since it makes refactoring way way easier.

timabbott replied:
``` diff
diff --git a/web/src/saved_snippets_ui.ts b/web/src/saved_snippets_ui.ts
index 86472fe2ed..a08e7f939d 100644
--- a/web/src/saved_snippets_ui.ts
+++ b/web/src/saved_snippets_ui.ts
@@ -18,16 +18,16 @@ import type {StateData} from "./state_data";
 let saved_snippet_dropdown_widget: dropdown_widget.DropdownWidget;
 
 function submit_create_saved_snippet_form(): void {
-    const title = $<HTMLInputElement>("#add-new-saved-snippet-modal .title").val()?.trim();
-    const content = $<HTMLInputElement>("#add-new-saved-snippet-modal .content").val()?.trim();
+    const title = $<HTMLInputElement>("#add-new-saved-snippet-modal .saved-snippet-title").val()?.trim();
+    const content = $<HTMLInputElement>("#add-new-saved-snippet-modal .saved-snippet-content").val()?.trim();
     if (title && content) {
         dialog_widget.submit_api_request(channel.post, "/json/saved_snippets", {title, content});
     }
 }
 
 function update_submit_button_state(): void {
-    const title = $<HTMLInputElement>("#add-new-saved-snippet-modal .title").val()?.trim();
-    const content = $<HTMLInputElement>("#add-new-saved-snippet-modal .content").val()?.trim();
+    const title = $<HTMLInputElement>("#add-new-saved-snippet-modal .saved-snippet-title").val()?.trim();
+    const content = $<HTMLInputElement>("#add-new-saved-snippet-modal .saved-snippet-content").val()?.trim();
     const $submit_button = $("#add-new-saved-snippet-modal .dialog_submit_button");
 
     $submit_button.prop("disabled", true);
@@ -89,7 +89,7 @@ function item_click_callback(
             form_id: "add-new-saved-snippet-form",
             update_submit_disabled_state_on_change: true,
             on_click: submit_create_saved_snippet_form,
-            on_shown: () => $("#saved-snippet-title").trigger("focus"),
+            on_shown: () => $("#add-saved-snippet-title").trigger("focus"),
             post_render: saved_snippet_modal_post_render,
         });
     } else {
diff --git a/web/styles/compose.css b/web/styles/compose.css
index 1fff79fbe5..1ec95b3081 100644
--- a/web/styles/compose.css
+++ b/web/styles/compose.css
@@ -1622,12 +1622,12 @@ textarea.new_message_textarea {
 }
 
 #add-new-saved-snippet-modal {
-    & .title {
+    & .saved-snippet-title {
         width: 97%;
         margin-bottom: 20px;
     }
 
-    & .content {
+    & .saved-snippet-content {
         width: 97%;
         resize: vertical;
     }
diff --git a/web/templates/add_saved_snippet_modal.hbs b/web/templates/add_saved_snippet_modal.hbs
index 7aa43f3bf5..75e73ef0b5 100644
--- a/web/templates/add_saved_snippet_modal.hbs
+++ b/web/templates/add_saved_snippet_modal.hbs
@@ -1,8 +1,8 @@
 <div id="add-new-saved-snippet-modal" class="new-style">
     <form id="add-new-saved-snippet-form" class="new-style">
         <label for="title" class="modal-field-label">{{t "Title" }}</label>
-        <input id="saved-snippet-title" type="text" name="title" class="modal_text_input title" value="" autocomplete="off" spellcheck="false" autofocus="autofocus"/>
+        <input id="new-saved-snippet-title" type="text" name="title" class="modal_text_input saved-snippet-title" value="" autocomplete="off" spellcheck="false" autofocus="autofocus"/>
         <div>{{t "Content" }}</div>
-        <textarea class="content settings_textarea" rows="4"></textarea>
+        <textarea class="settings_textarea saved-snippet-content" rows="4"></textarea>
     </form>
 </div>
```

timabbott commented:
Made the change noted above and marked to merge once CI passes. Several follow-ups have their own issues:
- #31827
- #31829
- #31830
